### PR TITLE
fix FuzzyQuery in keyword field will use IndexOrDocValuesQuery when both of index and doc_value are true

### LINKED ISSUE #14377: [BUG] FuzzyQuery in keyword type does not use the IndexOrDocValuesQuery when index is and doc_value is true
### Describe the bug

FuzzyQuery in keyword field does not use the `IndexOrDocValuesQuery` when index and doc_value are true.

In `FuzzyQueryBuilder.doToQuery()`,
https://github.com/opensearch-project/OpenSearch/blob/0d38d1498bc6c4b87461378ce91447dca46b2b58/server/src/main/java/org/opensearch/index/query/FuzzyQueryBuilder.java#L360
it calls `fieldType.fuzzyQuery()`, which will go into `StringFieldType.fuzzyQuery()`.
https://github.com/opensearch-project/OpenSearch/blob/0d38d1498bc6c4b87461378ce91447dca46b2b58/server/src/main/java/org/opensearch/index/mapper/StringFieldType.java#L82
not the `KeywordFieldMapper.fuzzyQuery()`, the parameters in `KeywordFieldMapper.fuzzyQuery()` is different with `FuzzyQueryBuilder.doToQuery()` calling.
https://github.com/opensearch-project/OpenSearch/blob/0d38d1498bc6c4b87461378ce91447dca46b2b58/server/src/main/java/org/opensearch/index/mapper/KeywordFieldMapper.java#L536

### Related component

Search

**Host/Environment (please complete the following information):**
 - Version os2.14.0

**Additional context**

### PR DESCRIPTION
**Pull request by kkewwei**
### Description
Fix FuzzyQuery in keyword field will use IndexOrDocValuesQuery when both of index and doc_value are true.

### Related Issues
Resolves #14377

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kkewwei** [comment]
@harshavamsi  @msfroh, please help review the code when you are free, I am very grateful for this.

File: server/src/main/java/org/opensearch/index/mapper/StringFieldType.java [hunk 1]
```diff
@@ -102,6 +103,35 @@ public Query fuzzyQuery(
...
+    public Query fuzzyQuery(
+        Object value,
+        Fuzziness fuzziness,
+        int prefixLength,
+        int maxExpansions,
+        boolean transpositions,
+        MultiTermQuery.RewriteMethod method,
+        QueryShardContext context
+    ) {
+        if (context.allowExpensiveQueries() == false) {
```
Comment (dblock) [inline server/src/main/java/org/opensearch/index/mapper/StringFieldType.java:116]: if (! ...)

**Review by dblock — CHANGES_REQUESTED** [review]
Thanks, see below.

**Review by harshavamsi — COMMENTED** [review]
Thanks for catching this @kkewwei! LGTM!

**Review by harshavamsi — APPROVED** [review]

**Comment by kkewwei** [comment]
> null



> ❌ Gradle check result for [751fc36](https://github.com/opensearch-project/OpenSearch/commit/751fc363102cab9fadb65ba014544c4df423361f): [null](https://build.ci.opensearch.org/job/gradle-check/41186/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

SearchRestCancellationIT.testAutomaticCancellationMultiSearchDuringQueryPhase #13674
RecoveryFromGatewayIT.testMultipleReplicaShardAssignmentWithDelayedAllocationAndDifferentNodeStartTimeInBatchMode #14346
RemoteSplitIndexIT.classMethod #13828

**Comment by kkewwei** [comment]
@dblock, please have s another look when you are free.

File: server/src/main/java/org/opensearch/index/mapper/StringFieldType.java [hunk 2]
```diff
@@ -102,6 +103,35 @@ public Query fuzzyQuery(
...
+        Object value,
+        Fuzziness fuzziness,
+        int prefixLength,
+        int maxExpansions,
+        boolean transpositions,
+        MultiTermQuery.RewriteMethod method,
+        QueryShardContext context
+    ) {
+        if (!context.allowExpensiveQueries()) {
+            throw new OpenSearchException(
```
Comment (dblock) [inline server/src/main/java/org/opensearch/index/mapper/StringFieldType.java:117]: This path still needs a test, or is there one I am missing?

**Review by dblock — DISMISSED** [review]
I'm not super familiar with this code, maybe @andrross or someone else is?

File: server/src/main/java/org/opensearch/index/mapper/StringFieldType.java (hunk 2 above)
Comment (harshavamsi) [reply server/src/main/java/org/opensearch/index/mapper/StringFieldType.java:117]: @dblock we do have tests already that catch this 

https://github.com/opensearch-project/OpenSearch/blob/3a0c0c0b38c0b42bc519c3673d5cd4a1e3379550/server/src/test/java/org/opensearch/index/mapper/KeywordFieldTypeTests.java#L291C1-L295C132

**Review by andrross — APPROVED** [review]

**Review by andrross — APPROVED** [review]

**Comment by kkewwei** [comment]
@andrross  please help merge, thank you.

**Comment by kkewwei** [comment]
> ❌ Gradle check result for [5e7ab2a](https://github.com/opensearch-project/OpenSearch/commit/5e7ab2afbdbe636bb43b523b684743a89f4d9e02): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/41523/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

SearchWithRandomExceptionsIT.testRandomExceptions #14144

**Review by andrross — APPROVED** [review]

**Pull request merged**